File: flutter/lib/main_screen.dart
// Copyright 2019 Google Inc. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import 'package:bikebuds/config.dart';
import 'package:bikebuds/main_content.dart';
import 'package:bikebuds/user_state.dart';
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'package:transparent_image/transparent_image.dart';
import 'package:url_launcher/url_launcher.dart';

import 'pages/pages.dart' as p;

class MainScreen extends StatefulWidget {
  @override
  _MainScreenState createState() => _MainScreenState();
}

class _MainScreenState extends State<MainScreen> {
  int _selectedDrawerItem = 0;

  @override
  Widget build(BuildContext context) {
    var userState = Provider.of<UserState>(context);
    var frontendUrl = Provider.of<Config>(context)?.config["devserver_url"];
    var pages = p.createPages(frontendUrl);
    return Scaffold(
      appBar: AppBar(
        title: Text("Bikebuds"),
      ),
      drawer: buildDrawer(pages, userState),
      body: MainContent(mainContentGlobalKey, pages, _selectedDrawerItem),
    );
  }

  Drawer buildDrawer(List<p.Page> pages, UserState userState) {
    var name = userState.displayName;
    var email = userState.email;
    var photoUrl = userState.photoUrl;
    var profilePhoto = photoUrl == null
        ? MemoryImage(
            kTransparentImage,
          )
        : NetworkImage(photoUrl);
    final List<Widget> children = [
      UserAccountsDrawerHeader(
          accountEmail: Text(email),
          accountName: Text(name),
          currentAccountPicture: CircleAvatar(
            backgroundImage: profilePhoto,
          ))
    ];
    for (int i = 0; i < pages.length; i++) {
      children.add(ListTile(
        title: Text(pages[i].title),
        onTap: () {
          setState(() {
            _selectedDrawerItem = i;
            Navigator.pop(context);
          });
        },
      ));
    }
    return Drawer(
      child: Column(
        children: <Widget>[
          Expanded(
            child: ListView(
              // Important: Remove any padding from the ListView.
              padding: EdgeInsets.zero,
              children: children,
            ),
          ),
          Container(
            child: Align(
              alignment: FractionalOffset.bottomCenter,
              child: Container(
                child: Column(
                  children: <Widget>[
                    Divider(),
                    AboutListTile(
                      child: Center(
                        child: Text('ToS - Privacy'),
                      ),
                      aboutBoxChildren: <Widget>[
                        ListTile(
                          title: Text("Terms of Service"),
                          onTap: () {
                            launch("https://bikebuds.com/tos");
                            Navigator.pop(context);
                          },
                        ),
                        ListTile(
                          title: Text("Privacy Policy"),
                          onTap: () {
                            launch("https://bikebuds.com/privacy");
                            Navigator.pop(context);
                          },
                        ),
                      ],
                    )
                  ],
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}

File: gae/frontend/src/MainScreen.js
/**
 * Copyright 2018 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import PropTypes from 'prop-types';
import React, { Component } from 'react';

import { createStyles, withStyles } from '@material-ui/core/styles';

import Chrome, { EmbedChrome } from './Chrome';
import FcmManager from './FcmManager';
import MainContent from './MainContent';
import ProfileWrapper, { ProfileState } from './ProfileWrapper';
import SwagWrapper from './SwagWrapper';

class MainScreen extends Component {
  static styles = theme =>
    createStyles({
      root: {
        display: 'flex',
        height: '100%',
        width: '100%',
      },
    });

  static propTypes = {
    embed: PropTypes.bool.isRequired,
    match: PropTypes.object.isRequired,
    firebase: PropTypes.object.isRequired,
    firebaseUser: PropTypes.object.isRequired,
    firebaseToken: PropTypes.string.isRequired,
  };

  constructor(props) {
    super(props);
    this.state = {
      bikebudsApi: null,
      profile: new ProfileState(this.handleProfileUpdated),
    };
  }

  handleSwagReady = client => {
    this.setState({
      bikebudsApi: client.apis.bikebuds,
      adminApi: client.apis.admin,
    });
  };

  handleProfileUpdated = profile => {
    this.setState({
      profile: profile,
    });
  };

  handleFcmMessage = payload => {
    console.log('Main.handleFcmMessage', payload);
  };

  render() {
    const mainContent = (
      <MainContent
        match={this.props.match}
        firebase={this.props.firebase}
        firebaseUser={this.props.firebaseUser}
        bikebudsApi={this.state.bikebudsApi}
        profile={this.state.profile}
      />
    );

    return (
      <div className={this.props.classes.root} data-testid="main">
        <SwagWrapper
          firebaseUser={this.props.firebaseUser}
          firebaseToken={this.props.firebaseToken}
          onReady={this.handleSwagReady}
        />
        {this.state.bikebudsApi && (
          <ProfileWrapper
            bikebudsApi={this.state.bikebudsApi}
            match={this.props.match}
            profileState={this.state.profile}
          />
        )}
        {!this.props.embed && this.state.bikebudsApi && (
          <FcmManager
            firebase={this.props.firebase}
            bikebudsApi={this.state.bikebudsApi}
            onMessage={this.handleFcmMessage}
          />
        )}
        {this.props.embed ? (
          <EmbedChrome
            history={this.props.history}
            profile={this.state.profile}
          >
            {mainContent}
          </EmbedChrome>
        ) : (
          <Chrome profile={this.state.profile}>{mainContent}</Chrome>
        )}
      </div>
    );
  }
}
export default withStyles(MainScreen.styles, { withTheme: true })(MainScreen);
